src/dataflow/src/render/sources.rs

```diff
@@ -809,8 +821,65 @@ where
     K2: Codec + timely::Data,
     V2: Codec + timely::Data,
 {
-    let sealed_stream =
-        stream.conditional_seal(&source_name, bindings_stream, write, bindings_write);
+    let sealed_stream = stream.conditional_seal(
+        &source_name,
+        bindings_stream,
+        write.clone(),
+        bindings_write.clone(),
+    );
+
+    // The compaction since of persisted sources is upper-bounded by two frontiers: a) the
+    // compaction that the coordinator allows, and b) the seal frontier of the involved persistent
+    // collections.
+
+    // We need to obey a) because that's what the coordinator wants, and b) because if we persisted
+    // past (or up to) the seal frontier we would not be able to distinguish between valid and
+    // invalid updates based on the upper seal frontier when restarting.
+    //
+    // We get the upper seal frontier using a probe handle, and we expect the coordinator to call
+    // `allow_compaction` on the stashed `CompactionHandle` to present it's view of what should be
+    // allowed.
+
+    let mut probe: ProbeHandle<Timestamp> = ProbeHandle::new();
+    sealed_stream.probe_with(&mut probe);
+    let upper_compaction_bound_fn = move || {
+        probe.with_frontier(|frontier| {
+            Antichain::from(
+                frontier
+                    .iter()
+                    .cloned()
+                    // Is is **very** important to not compact up to the seal frontier, as
+                    // mentioned above.
+                    .map(|e| e.saturating_sub(1))
+                    .collect::<Vec<_>>(),
+            )
+        })
+    };
+
+    let mut compaction_handle = AtomicCompactionHandle::new(&write);
+    compaction_handle.add_stream(&bindings_write);
+
+    let compaction_handle =
+        BoundedCompactionHandle::new(upper_compaction_bound_fn, compaction_handle);
+
+    let compaction_token: Rc<Box<dyn CompactionHandle>> = Rc::new(Box::new(compaction_handle));
+    additional_tokens.push(Rc::clone(&compaction_token) as Rc<dyn Any>);
+
+    let previous = storage_state
+        .compaction_handles
+        .insert(source_id.clone(), Rc::downgrade(&compaction_token));
+
+    match previous {
+        Some(compaction_handle) => {
```

Comment:
spitballing: we could sidestep all of this by storing something like `HashMap<GlobalId, Vec<(CompactionHandle, Weak<ProbeHandle>>>>`

and then read out the sealed frontier for each stream individually, and compact the stream up to min(allowed_compaction, seal_frontier - 1)`

Reply:
We want to intentionally fail because there must only be one rendered instance of a source. Plus, I like that the code that handles `AllowSourceCompaction` only has this one `allow_compaction()` call to make without worrying about the details of `ProbeHandles` and whatnot. WDYT?

Reply:
yeah, i def agree with aljoscha here